### ISSUE TITLE
RFC: Add complex exponential exp(z) function

### ISSUE THREAD

**Issue by rreusser** [stdlib-js/stdlib#145]
## Checklist

> Please ensure the following tasks are completed before filing an issue.

-   [x] Read and understood the [Code of Conduct][code-of-conduct].
-   [x] Searched for existing issues and pull requests.
-   [x] If this is a general question, searched the [FAQ][faq] for an existing answer.
-   [x] If this is a feature request, the issue name begins with `RFC:`.

## Description

> Description of the issue (or feature request).

This function is *very* similar to `cis(z)`. See #143. It has just a bit of special case handling to avoid unnecessary evaluations in degenerate cases (nan/inf). I suspect perhaps the right answers happen automatically based on IEEE 754 rules and that the extra logic just short-circuits some evaluations.

## Questions

> Any questions for reviewers?

No.

## Other

> Any other information relevant to this issue (or feature request)? This may include screenshots, references, stack traces, sample output, and/or implementation notes.

- [julia: exp(z)](https://github.com/JuliaLang/julia/blob/0df3f7dc0280525349ae23e3ac4fc96621befc50/base/complex.jl#L551)

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[faq]: https://github.com/stdlib-js/stdlib/blob/develop/FAQ.md

**Comment by rreusser** [comment]
My trivial implemenation has been done for a long time, but for my own sake, teasing apart corner cases from [julia's implementation](https://github.com/JuliaLang/julia/blob/9f8bdf459ea04f5a91dae1083ffca4a0f7da7d18/base/complex.jl#L567):

- Real part NaN
  - [x] `cexp(NaN, 0) -> Complex(NaN, 0)` (python agrees on this one)
  - [x] `cexp(NaN, im) -> Complex(NaN, NaN)`
- Imaginary part not finite
  - [ ] `cexp(+Inf, +/-INF) -> Complex(-Inf, NaN)`
    - seems arbitrary
  - [ ] `cexp(-Inf, +/-INF) -> Complex(-0, 0 * sign(im))`
    - also seems arbitrary. what does im(z) say about sign(sin(im)) if it's infinite?
  - [x] `cexp(NaN, +/-INF) -> Complex(NaN, NaN)`
- Regular re/im
  - [x] `cexp(re, 0) -> Complex(exp(re), im)`
  - [x] `cexp(re, im) -> Complex(exp(re) * cos(im), exp(re) * sin(im))`

@kgryte My current perhaps-naive implementation returns `[NaN, NaN]` for all of the above unchecked cases. Python similarly returns `[NaN, NaN]`.

For the infinite cases, they seem to be doing some sign-copying that I can't quite figure out the value of. `sign(sin(x)) != sign(x)`, especially when x is infinite, so I'm not sure why they're copying signs.

Bottom line, no deep exploration required; I'm submitting a PR currently and just wanted to document some of the corner cases here.

**Comment by kgryte** [comment]
My guess is that they are being sensitive to branch cuts and may also be attempting to ensure consistent behavior with `cpow`. From a superficial scan of the Kahan paper, not immediately evident the rationale, but may be an extension of the generalization of `x^y` to complex numbers.

References:

-   Julia (blame): https://github.com/JuliaLang/julia/blame/9f8bdf459ea04f5a91dae1083ffca4a0f7da7d18/base/complex.jl#L567-L588
-   Julia (PR): 2891
-   Kahan: https://people.freebsd.org/~das/kahan86branch.pdf
-   Python: https://github.com/python/cpython/blob/42c35d9c0c8175332f50fbe034a001fe52f057b9/Modules/cmathmodule.c#L546
-   C: http://en.cppreference.com/w/c/numeric/complex/cexp
